firebase-perf/src/test/java/com/google/firebase/perf/application/FrameMetricsRecorderTest.java

```diff
@@ -0,0 +1,405 @@
+    ArgumentCaptor<Activity> activityCaptor = ArgumentCaptor.forClass(Activity.class);
+    recorder.start();
+    recorder.stop();
+    recorder.start();
+    recorder.stop();
+    verify(fma, times(2)).add(activityCaptor.capture());
+    verify(fma, times(2)).remove(activityCaptor.capture());
+    List<Activity> activities = activityCaptor.getAllValues();
+    Assert.assertTrue(activities.stream().allMatch(a -> a == activity));
+  }
+
+  @Test
+  public void start_callsFrameMetricsAggregator() {
+    verify(fma, times(0)).add(nullable(Activity.class));
+    recorder.start();
+    verify(fma, times(1)).add(nullable(Activity.class));
+  }
+
+  @Test
+  public void start_whileAlreadyStarted_fails() {
+    recorder.start();
+    recorder.start();
+    verify(fma, times(1)).add(nullable(Activity.class));
+  }
+
+  @Test
+  public void start_afterPreviousEnded_succeeds() {
+    stubSnapshotToDoNothing();
+    recorder.start();
+    recorder.stop();
+    recorder.start();
+    verify(fma, times(2)).add(nullable(Activity.class));
+  }
+
+  @Test
+  public void stop_callsFrameMetricsAggregator() {
+    stubSnapshotToDoNothing();
+    recorder.start();
+    recorder.stop();
+    verify(fma, times(1)).remove(nullable(Activity.class));
+    verify(fma, times(1)).reset();
+  }
+
+  @Test
+  public void stop_whileNotStarted_fails() {
+    stubSnapshotToDoNothing();
+    recorder.stop();
+    verify(fma, times(0)).remove(nullable(Activity.class));
+    verify(fma, times(0)).reset();
+    recorder.start();
+    recorder.stop();
+    recorder.stop();
+    verify(fma, times(1)).remove(nullable(Activity.class));
+    verify(fma, times(1)).reset();
+  }
+
+  @Test
+  public void stop_snapshotsFrameMetricsAggregator_beforeReset() {
+    stubSnapshotToDoNothing();
+    recorder.start();
+    recorder.stop();
+    InOrder orderVerifier = inOrder(recorder, fma);
+    orderVerifier.verify(recorder).snapshot();
+    orderVerifier.verify(fma).reset();
+  }
+
+  @Test
+  public void startSubTrace_whenNotRecording_fails() {
+    stubSnapshotToDoNothing();
+    Object uiState = new Object();
+    recorder.startSubTrace(uiState);
+    verify(subTraceMap, times(0)).put(nullable(Object.class), nullable(PerfFrameMetrics.class));
+
+    recorder.start();
+    recorder.stop();
+    recorder.startSubTrace(uiState);
+    verify(subTraceMap, times(0)).put(nullable(Object.class), nullable(PerfFrameMetrics.class));
+  }
+
+  @Test
+  public void startSubTrace_whenSameSubTraceWithGivenKeyIsAlreadyOngoing_fails() {
+    doReturn(Optional.of(frameMetrics1)).when(recorder).snapshot();
+    ArgumentCaptor<Object> objectCaptor = ArgumentCaptor.forClass(Object.class);
+    Object uiState1 = new Object();
+    Object uiState2 = new Object();
+
+    recorder.start();
+    recorder.startSubTrace(uiState1);
+    verify(subTraceMap, times(1)).put(objectCaptor.capture(), nullable(PerfFrameMetrics.class));
+    Assert.assertSame(frameMetrics1, subTraceMap.get(uiState1));
```

Comment:
Use a static import.

Reply:
Could you elaborate? I'm not sure what you mean in this context.